tests/gen/gen_auto.go

```diff
@@ -0,0 +1,245 @@
+		configurator:     configurator,
+		generatedDocKeys: make(map[string][]string),
+	}
+}
+
+type randomDocGenerator struct {
+	configurator docsGenConfigurator
+
+	generatedDocKeys map[string][]string
+	random           rand.Rand
+}
+
+func (g *randomDocGenerator) generateDocs(options ...Option) ([]GeneratedDoc, error) {
+	err := g.configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *g.configurator.random
+
+	resultDocs := make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(g.configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.configurator.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			resultDocs = append(resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.configurator.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.configurator.usageCounter.getNextTypeIndForField(secondaryType, field)
+	docKey := g.generatedDocKeys[field.Schema][ind]
+	return docKey
+}
+
+func (g *randomDocGenerator) getDocKey(typeDef *client.CollectionDefinition, doc map[string]any) (string, error) {
+	clientDoc, err := client.NewDocFromJSON([]byte(createDocJSON(typeDef, doc)))
+	if err != nil {
+		return "", err
+	}
+	return clientDoc.Key().String(), nil
+}
+
+func (g *randomDocGenerator) generateRandomDocs(order []string) ([]DocsList, error) {
+	result := []DocsList{}
+	for _, typeName := range order {
+		col := DocsList{ColName: typeName}
+		typeDef := g.configurator.types[typeName]
+
+		currentTypeDemand := g.configurator.docsDemand[typeName]
+		averageDemand := currentTypeDemand.getAverage()
+		for i := 0; i < averageDemand; i++ {
+			newDoc := make(map[string]any)
+			for _, field := range typeDef.Schema.Fields {
+				if field.Name == request.KeyFieldName {
+					continue
+				}
+				if field.IsRelation() {
+					if field.IsPrimaryRelation() {
+						newDoc[field.Name+request.RelatedObjectID] = g.getNextPrimaryDocKey(typeName, &field)
+					}
+				} else {
+					fieldConf := g.configurator.config.ForField(typeName, field.Name)
+					newDoc[field.Name] = g.generateRandomValue(typeName, field.Kind, fieldConf)
+				}
+			}
+			docKey, err := g.getDocKey(&typeDef, newDoc)
+			if err != nil {
+				return nil, err
+			}
+			g.generatedDocKeys[typeName] = append(g.generatedDocKeys[typeName], docKey)
+			col.Docs = append(col.Docs, newDoc)
+		}
+		result = append(result, col)
+	}
+	return result, nil
+}
+
+func getRandomString(random *rand.Rand, n int) string {
```

Comment:
todo: There appears to be no protection against collisions here, which exposes us to flaky tests.

Consider the schema:
```
type User {
   Name: String # max: 1 // I cant remember the exact syntax, is not important here
}
```
1 in 26*2 docs would be identical and would cause tests to fail if they try to insert them.  Yes the odds increase with the number of chars, but we run the CI workflow a lot, and each run runs all the tests ~20 times, and there will hopefully be more than just one test using this doc generation, so I'd still expect us to be impacted by this if all the tests had 6 or so char strings.

I think it would be worth ensuring that each call to `generateDocs` cannot return the same string twice, for at least short-medium length strings.

This problem will exist for other field types too. although as we don't support `int8` etc it is unlikely to be a real problem there atm (maybe for bool inline-arrays).

Reply:
the generator doesn't support at the moment unique fields. The example you provided will just generate documents the have the same values for `Name` field.
I plan on extending the generators capabilities like:
```graphql
type User {
   Name: String # min: 1, max: 1, unique
}
```
And a corresponding new option `WithUniqueField("User", "Name")`
But this will be probable another PR

Reply:
I don't understand - you want to merge this knowing that it will cause our tests to be flaky?

Reply:
Just to clarify why this is a problem - you cannot create two documents in a defra db with the same field values - it would result in the same doc key and the second create command will result in a `document already exists` error

Reply:
I'm aware of this constrain. That's why it's a bit confusing when you talk about the field uniqueness rather than doc uniqueness.

I've created a task for field uniqueness https://github.com/sourcenetwork/defradb/issues/2069.

For document uniqueness the story is a bit different. There are different ways of solving it.

One of them is to store a map of all dockeys generated and make sure newly created doc doesn't have the same dockey, and if so we should generated it again until the dockey is unique. The problem with this is that if the number of available permutations is close to number of demanded docs the generator will have hard times picking a unique combination. We could mitigate it by running some kind analysis after doc demand calculation to determine how likely we to collide.

I have some other interesting options in mind but I don't think it's worth the trouble. All this additional bookkeeping will make the code much more complex and less performant.

I think if the client wants to make sure there is no (or almost no) collision, he should provide a reasonable setup.
If his SDL is 
```graphql
type User {
    name: String # len: 1
}
```
and her requests 100 docs to generate, he is intentionally looking for trouble.

You said the if we use any but 8-bit integer it should be fine, but with string length of 3 you get twice as many permutations as with 16-bit integer.

Reply:
The problem is that the risk on collision needs to be very low for it to not annoy us by tripping up the CI every now and then.  Say we have 50 tests that use this, and they each generate 11 docs, each push will run that 20 odd times (os, mutation type, client type, etc), and maybe on average each PR will have the CI run 20 times.  That is roughly 50*10*20*20=200000 chances per PR for duplicate docs to cause the CI to fail - increasing as this tool gets used by more tests, and if/as we hire more devs (more PRs).

And we can avoid this by just generating sequential values, or tracking which field values have been created in a test (very cheap, so long as the test is not creating billions of documents).

Reply:
I got too curious and checked the odds, if the maths in `50*(11-1)*(11-1)*20*20` is correct, for a schema with a single int32 field (e.g. Age), the odds of the tests not flaking out in 1 PR (20 pushes, 50 tests, 11 docs per test), are `0.999069110971`, i.e. 1 in 1000 chance of failing.  If we have 10 PRs, the odds of no failure = `0.990730008019` (1 in 100).  100 PRs = `0.911072996106` (9 in 100), 1000 PRs = `0.39403222849` (6 out of 10 chance of failure).  The odds are worse for short strings.

Reply:
I've reviewed the calculations you provided regarding the probabilities of collisions in our CI tests. I noticed a small discrepancy in the multiplication that affects the final number of opportunities for collision.

> if the maths in 50*(11-1)*(11-1)*20*20 is correct

Seems like you incorrectly took a square of number of tests. It should be just `50*(11-1)*20*20`
Which gives us 200,000 instead of 2,000,000 chances for duplication.

With the corrected calculation, the probability of at least one collision happening across 100 PRs, when using a 5-character string, is approximately 5.12%.

To be honest, I'm not sure why we are so much concerned about this kind of flakiness. I can hardly remember any of my PRs that didn't fails for a not legit reason (changes-related), so I just had to restart the job.

I don't say that we should neglect the number (your or mine) but I'd say it's not that high of a priority now. I propose to add a task for different way of random value generation. Maybe with lexicographically incrementable strings as the default option.

Reply:
Ah, I think it is (11-1)^2 because each document (n-1) has n-1 chances to collide with another doc.

>  I propose to add a task for different way of random value generation. Maybe with lexicographically incrementable strings as the default option.

I'm a bit uncomfortable with that, but can accept it - it isn't like this PR converts all our tests to using it, and it should be fairly quick to fix if/when it starts flaking out (although that will cause delay and annoyance when working on unrelated stuff).

Reply:
@fredcarle Tagging you here to get a third opinion on this conversation

Reply:
I think that if we setup our doc generation properly we should have minimal impact due to collisions. My opinion is to give this a try and fix it if it becomes a problem.

Reply:
added a task https://github.com/sourcenetwork/defradb/issues/2072